network/dhcp-client: Allow saving dhcp4 leases in persistent directory and include config options to use them

rajibade opened the pull request:
Hi,
This is just a Draft PR for the introduction of the functionality discussed in #37876. This does a few things
1. Implements an opt-in option for persistently storing leases under dhcp
2. Saves acquired leases using already implemented functions
3. tries to load back and reuse the saved leases

There have been a few pitfalls to when writing this. I have a thread up in the mailing list to talk about the technical issues encountered. But for this draft, i would like some feedback on if this is viable for an addition or I need to rework some things. I did a lot of manual testing for this, But I will be writing unit tests if this approach is okay.

Thanks

rajibade committed:
network/dhcp-client: Expand network config option for client persistent leases

rajibade committed:
network/dhcp-client: Add /var persistent storage creation

rajibade committed:
network/dhcp-client: Fix paths creation and writing logic

rajibade committed:
network/dhcp-client: Load saved lease w/ timestamp inclusion

rajibade committed:
network/dhcp-client: Add comments/remove debug prints

rajibade committed:
network/dhcp-client: Saving UTC timestamp for lease time tracking

rajibade committed:
network/dhcp-client: Fix missing include for lease save/load

rajibade reviewed (commented):
Missing include. fixed with new commit

rajibade commented:
Thanks for the comment. Sorry I haven't had time for an update. Getting the boot times in the files was trivial. I'm having a bit of a issue understanding `vl_method_set_persistent_storage()`. I see it becomes active when the external service `systemd-networkd-persistent-storage.service` is run, which puts up a `Ready` signal on the Varlink. does the file descriptor `fd` need to be passed into relevant functions to access the directory? or does it poll continually when access is requested? I am really asking this because the persistent storage is enabled when the above service is run and it stays on till systemd shutdown

yuwata commented:
The mechanism is required because when /var/ is in a separated filesystem from / or /usr/. In that case, networkd may be started before /var/ is mounted, and networkd cannot/should not write leases in /var/ before it is mounted.

See vl_method_set_persistent_storage(), and call a new function that enable saving aquired lease to the persistent storage, like we already do in manager_toggle_dhcp4_server_state() for DHCP server rather than client.

The helper service systemd-networkd-persistent-storage.service is typically started when /var/ is mounted (or started immediately if /var/ is not a mountpoing) during boot, and stopped before /var/ is unmounted. Of course, we can manually start/stop/restart the service, or may be stopped when /var/ is unmounted anytime.

So, I think it is better to postpone starting DHCP client until the helper service is started, and when it is started, read the saved lease, assign address and friends if it is still valid, rebind the lease for safety, and so on.

rajibade committed:
network/dhcp-client: Fix persistent storage writing with new functions and attempt lease validation

rajibade commented:
@yuwata Just put up an update to this. I was able to get the persistent storage working and in line with the helper service. Thanks for the hint. The loading works now and the address requests and routes function is called but there is a problem with validating the lease timing and assigning the correct timing so the system can continue to process the loaded lease. Please let me know if I am overthinking the following. 

1. I save the lease file with the CLOCK_REALTIME AND CLOCK_BOOTTIME, and BOOT_ID, which is used to track when the lease was acquired and later how long has expired since.

2. When loading the lease back, the previous times are stored in the lease structure using dhcp_lease_set_timestamp, do I need to do this? I assume I will need this for comparing times down the line.

3. During the lease validation, based on whether the boot IDs are the same, I check if (saved_time + lease_lifetime) > current_time, either boot or real time. if this is correct, then the lease is invalid correct?

4. But what if the lease is still valid? I tried to use triple_timestamp_from_realtime(since this has map_clock_usec_raw) to get the correct boot time back from the old realtime, then set the new lease timestamp as this is what would be used for processing the timings right? but I think this is failing to properly register. Am I missing something in terms of re-assigning the times back? also is there any other difference in the process between a full reboot vs userspace reboot(systemctl soft-reboot/restarting networkd), as I am seeing that I could just use the realtime as the differentiator in both cases, instead of relying on boottime? 

Please let me know what you think of this. Look at dhcp4_validate_persistent_lease() function in the PR for context of what I have done. I left some commented out code for some more context.

rajibade commented:
Nvm. Figured it out :)

rajibade committed:
network/dhcp-client: Fix correct time left on persistent loaded lease and lease loading procedure on full reboot

rajibade committed:
network/dhcp-client: Add option to use an expired lease and allocate updated lease file to client structure

rajibade committed:
network/dhcp-client: Cleanup code and extra /var directories. Fix empty config option check

rajibade commented:
Hi @yuwata @poettering Can I get a review for this? This is a working implementation but i'd like some feedback if there's anything to change here. Some descriptions here about what's going on in the code and my thoughts

1. I was able to get the persistent writing in sync with when the varlink method put up the ready signal. From your earlier comment 
> The mechanism is required because when /var/ is in a separated filesystem from / or /usr/. In that case, networkd may be started before /var/ is mounted, and networkd cannot/should not write leases in /var/ before it is mounted.

I added a flag for readiness to the manager to propagate that signal, in case the client is about to start before the storage is ready and restart the client in `dhcp4_start`. I'm not exactly sure about timings but this was the best I could make of this. It also guarantees since the client needs to stop first, if anything happens, it has the chance to start up again.

2. When I call `dhcp4_request_address_and_routes`, the annouce variable is `true`, is this fine?
3. For calculating the expiring time, I only used `CLOCK_REALTIME` saved in the lease file. I removed `BOOT_ID` but I kept `CLOCK_BOOTTIME` but I'll remove it since I don't think it's necessary. The logic for this is pretty straightforward but I included the function `sd_dhcp_client_update_lease_lifetime`, to reset the lease lifetime to however much is left since restarting and recalculate the timeout values on the client structure. 

Finally, I have included an option in addition to the persistent use of leases. It is essentially the option to use an expired lease. From the comments in the original issue, I know there was a bit of pushback with this but I do think this is fundamental especially if this persistent lease use is an included function. I have made these changes with the idea that its fully OPT-IN, including using the persistent leases without the expiry option. It will not interfere with normal DHCP operation of networkd when disabled. The reasons for this inclusion are as follows

1. If end user is using the persistent lease option, the user is somewhat aware of their network configuration and is generally fine persisting leases. If this is the case, they will be persisting a valid lease which could mean that there is little cause for conflicts. Normal persistent leases will expire as normal and will lose connection if server is not responsive. 
2. With the expiry option, end user gets the benefit of the persistent lease and the additional case of continuing to use the lease after expiration when the server is not responsive. I think this is fine in a known network config scenario. IP allocations isn't a problem if server is unresponsive anyway so all clients can continue to use their lease as normal. Scenario below

`Lease -> persist -> use during reboots -> expires = mitigated downtime using lease while rebooting and server inactive`
`Lease -> persist -> use during reboots -> expires -> extend expiration = mitigated downtime using lease while rebooting and server inactive, then mitigated downtime when lease expired and server inactive`

The implementation for that last scenario is still a bit unrefined. Currently I have the leases being extended every minute. This guarantees that when the server becomes responsive, there is little downtime before the client tries to reach out for a new lease. Problem is this causes an issue at scale, if multiple clients try to request at the same time, congestion will occur. 

So my additional proposal here is that, if using this expired lease scenario is fine, Randomise the extension value of the lease lifetime when expired (i.e some kind of fuzzing) so that not all clients will renew and hence request at the same rate. Can a value be left up to the end user to determine how much they would want? I plan on using some kind of upper bound value that the user can set, then the extension value can be randomised up to that value? Is this an inclusion that would be fine?

I think this would be a beneficial addition as a whole. Please let me know what you think. Thanks for reviewing.

yuwata commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -712,8 +726,12 @@ static int client_initialize(sd_dhcp_client *client) {
         client_set_state(client, DHCP_STATE_STOPPED);
         client->xid = 0;
 
+
```
Please do not insert unnecessary empty line.

yuwata commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -696,14 +699,25 @@ static int client_notify(sd_dhcp_client *client, int event) {
         client->fd = safe_close(client->fd);
 
+        if (client->lease)
+                saved_lease = sd_dhcp_lease_ref(client->lease);
```
`xyz_ref()` works fine with NULL, hence it is not necessary to conditionalize:
```c
_cleanup_(sd_dhcp_lease_unrefp) sd_dhcp_lease *saved_lease = sd_dhcp_lease_ref(client->lease);
```

rajibade replied:
Don't need it anymore :)

yuwata commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -2589,3 +2614,60 @@ static const char* const dhcp_state_table[_DHCP_STATE_MAX] = {
+        if (!path_is_safe(path))
+                return -EINVAL;
+
+        _cleanup_close_ int fd = AT_FDCWD;
```
Please add comment about why `AT_FDCWD` is set. See `sd_dhcp_server_set_lease_file()`.

yuwata commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -2589,3 +2614,60 @@ static const char* const dhcp_state_table[_DHCP_STATE_MAX] = {
+                return -EINVAL;
+
+        _cleanup_close_ int fd = AT_FDCWD;
+
```
Please drop unnecessary empty line.

yuwata commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -2589,3 +2614,60 @@ static const char* const dhcp_state_table[_DHCP_STATE_MAX] = {
+        _cleanup_close_ int fd = AT_FDCWD;
+
+        if (dir_fd >= 0) {
+                fd = fd_reopen (dir_fd, O_CLOEXEC | O_DIRECTORY | O_PATH);
```
Please drop unnecessary space before `(`

yuwata commented on src/libsystemd-network/sd-dhcp-lease.c:
```diff
@@ -1648,6 +1662,20 @@ int dhcp_lease_load(sd_dhcp_lease **ret, const char *lease_file) {
                         log_debug_errno(r, "Failed to parse T2 %s, ignoring: %m", t2);
         }
 
+        if (saved_realtime && saved_boottime) {
```
Then, if we read BOOTTIME, check the BOOTID, and if it matches the current boot id, then set all timestamps from the boottime.

If BOOTID does not match the current boot id, ignore BOOTTIME, and set timestamp from the realtime.

If both realtime and boottime is not set, maybe better to discard the saved lease, as we do not know if it is still valid.

yuwata commented on src/libsystemd-network/sd-dhcp-lease.c:
```diff
@@ -1648,6 +1662,20 @@ int dhcp_lease_load(sd_dhcp_lease **ret, const char *lease_file) {
+                triple_timestamp ts = {};
+                r = parse_timestamp(saved_realtime, &timestamp_usec);
+                if (r >= 0)
+                        ts.realtime = timestamp_usec;
```
Please use triple_timestamp_from_realtime()

yuwata commented on src/libsystemd-network/sd-dhcp-lease.c:
```diff
@@ -1648,6 +1662,20 @@ int dhcp_lease_load(sd_dhcp_lease **ret, const char *lease_file) {
+                if (r >= 0)
+                        ts.boottime = timestamp_usec;
+
+                dhcp_lease_set_timestamp(lease, &ts);
```
Not necessary.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+        }
+
+        if (dir_fd >= 0) {
+                lease_path = path_join("/var/lib/systemd/network", lease_file);
```
Please suffix with `/` for the directory, even `path_join()` gracefully handle that.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+
+        /* check if persistent lease is expired */
+        r = dhcp4_validate_persistent_lease(link, lease);
+        if (r < 0) {
```
Please drop unnecessary `{}`

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+        link->dhcp_lease = TAKE_PTR(lease);
+        link_dirty(link);
+
+        log_link_info(link, "Loaded saved DHCP lease");
```
Please append full stop to the log message.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+        if (link->manager->persistent_storage_fd < 0)
+                return -EBUSY;
+
+        _cleanup_free_ char *p = path_join("leases", link->ifname);
```
I'd like to name this as "dhcp-client-leases" or so, to make it consistent with dhcp-server-leases, and also for DHCPv6 leases in the future.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+}
+
+static int dhcp_configure_with_persistent_lease(Link* link) {
+
```
Please drop the empty line.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+        return 1;
+}
+
+static int dhcp_configure_with_persistent_lease(Link* link) {
```
Missing assertion for `link`

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+        if (r < 0) {
+                log_link_debug_errno(link, r, "Failed to get persistent lease path, ignoring: %m");
+                return r;
+        }
```
Same comment in the above.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+                /*load persistent lease, before restarting client */
+                r = dhcp4_load_persistent_lease(link);
+
+                if (r > 0) {
```
Also here. Please drop an empty line.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+        if (r > 0) {
+                r = sd_dhcp_client_set_lease_file(link->dhcp_client, dir_fd, lease_file);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to set lease file: %m");
```
Why error is not propagated?

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+
+                        r = dhcp4_request_address_and_routes(link, true);
+                        if (r < 0)
+                                log_link_warning_errno(link, r, "Failed to configure loaded lease: %m");
```
Also here.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+                }
+        }
+
+        return 1;
```
Please use 0 unless any specific means.

rajibade replied:
Need 1 here. I return 0 if there is no lease being used, but no erros

yuwata reviewed (changes requested):
Not fully reviewed. But already meany many comments.
Please follow our coding style. See https://systemd.io/CODING_STYLE/

rajibade commented:
Thanks for the comments. I'll take a look and fix them. Sorry for pinging and not putting it in review mode. I will do that next time.

rajibade commented on src/libsystemd-network/sd-dhcp-lease.c:
```diff
@@ -1354,6 +1356,14 @@ int dhcp_lease_save(sd_dhcp_lease *lease, const char *lease_file) {
+
+        r = sd_dhcp_lease_get_timestamp(lease, CLOCK_BOOTTIME, &t);
+        if (r >= 0)
+                fprintf(f, "TIMESTAMP_BOOTTIME=%s\n", FORMAT_TIMESTAMP_STYLE(t, TIMESTAMP_US));
```
Not Saving BOOTTIME anymore, will remove.

rajibade commented on src/network/networkd-dhcp4.c:
```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+        return 1;
+}
+
+static int dhcp4_load_persistent_lease(Link *link) {
```
saved sounds good. Will adjust log comments as well

rajibade commented on src/network/networkd-dhcp4.c:
```diff
@@ -1228,6 +1444,25 @@ static int dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
+                                 * TODO: Add fuzzing/randomness if required for this extension value. */
+                                usec_t extension = 1 * USEC_PER_MINUTE;
+
+                                r = sd_dhcp_client_update_lease_lifetime(client, link->dhcp_lease, extension);
```
> The implementation for that last scenario is still a bit unrefined. Currently I have the leases being extended every minute. This guarantees that when the server becomes responsive, there is little downtime before the client tries to reach out for a new lease. Problem is this causes an issue at scale, if multiple clients try to request at the same time, congestion will occur.

> So my additional proposal here is that, if using this expired lease scenario is fine, Randomise the extension value of the lease lifetime when expired (i.e some kind of fuzzing) so that not all clients will renew and hence request at the same rate. Can a value be left up to the end user to determine how much they would want? I plan on using some kind of upper bound value that the user can set, then the extension value can be randomised up to that value? Is this an inclusion that would be fine?

> I think this would be a beneficial addition as a whole. Please let me know what you think. Thanks for reviewing.

Please see other comment. Since expiry event is triggered, if user has option to extend expired lease, this will revalue lease with new lifetime value and update the timestamp. Please see `sd_dhcp_client_update_lease_lifetime()` I implemented in `sd-dhcp-client.c`. This scenario only comes about if saved lease expires.

rajibade commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -712,8 +726,12 @@ static int client_initialize(sd_dhcp_client *client) {
         client->lease = sd_dhcp_lease_unref(client->lease);
 
+        if (saved_lease)
+                client->lease = TAKE_PTR(saved_lease);
```
Was keeping convention in case persistent lease option was not used. I will test and see if it can be dropped safely

rajibade replied:
Not needed anymore. I will transition to BOUND state once lease is loaded, this will prevent the the entire client from being re-initialised and clearing the timers.

rajibade commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -2589,3 +2614,60 @@ static const char* const dhcp_state_table[_DHCP_STATE_MAX] = {
+                        return fd;
+        }
+
+        close_and_replace(client->lease_dir_fd, fd);
```
Will fix. used calling func

rajibade committed:
network/dhcp-client: Several code cleanups and functional comments

rajibade committed:
network/dhcp-client: Change client state when lease loaded and make sure config option is invalid if empty

rajibade committed:
network/dhcp-client: Initialise client lease dir fd

rajibade commented:
Latest changes:

- Reverted the client_initialize function back to original implementation. Why? Instead of stopping the function from resetting the timing event's when using loaded lease, just transition the client into bound state, then the client wouldn't have to re-initialize. 
- Removed BOOTTIME from lease file. Its not required for lease timing calculation. REALTIME will provide an absolute timing for when the lease is acquired(even during same boot period/ID)
- Handling the case where REALTIME is not available in the leasefile. If the option to persist is enabled, the client will fail to load the lease, however if the option to persist expiry, propagating the lease seems sufficient since the lease time will be limited either ways
- Code/cleanup

WIP
- Fuzzing expiry option lease extension value, due to possible network congestion problem

rajibade committed:
network/dhcp-client: Expand expired lease option, allow user configuration

rajibade commented:
Latest changes:

As explained at the end of this earlier [comment](https://github.com/systemd/systemd/pull/38806#issuecomment-3483041164), I believe that there is merit in allowing the usage of expired leases, especially in monitored environment and I have made some changes to reflect that here. 

- In addition to being able to use a saved lease, users can indicate to use a lease even if it is expired. Note this does not affect normal DHCP operations. If you have an active server, the leases will renew like normal, and if not then you are allowed to use your lease as a failsafe
- In order to make it as efficient as possible, there is a limit to the lifetime of the lease. I have made the conscious decision of making that 240s. This is based on general infra statistics I have and a user intuitive value. The reason for making it this small as well is from an optimal perspective. The expired leases will continue to be renewed internally(as long as no ack received), but will need a short time to gain a new lease once it can. This should prevent some level of ip conflicts that might linger for too long if the lifetime is too high, so small. 
- To that effect, there are now two addition user options that work with this Main persistent one, a value for the lifetime of the lease and a fuzz value(how much to randomise the lease lifetime within a range).
- The fuzz value is important in scenarios where there are multiple clients using this config option. It prevents all clients from having the same lease lifetime value, hence from all trying to renew/request at the same time which can lead to some network congestion is the scale is big enough. 
- Both options have default values, so if there is no fuzz, then same lease time gets used always(either internal default or specified), and if only fuzz, then default gets randomised. Fuzz percentage capped at 50%.

rajibade commented:
Poke @yuwata. please also look at the above comments for detailed explanations.

rajibade commented:
Hi @yuwata, is there going to be any forward consideration into this pr?

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -946,11 +947,28 @@ static int link_save(Link *link) {
+
+                        if (r == -EBUSY) {
+                                /* not ready */
+                        } else if (r < 0) {
```
```c
if (r < 0 && r != -EBUSY)  /* EBUSY: not ready */
        log_link_debug_errno(...);
else if (r > 0) {
        ...;
}
```

yuwata commented on src/network/networkd-network.c:
```diff
@@ -1133,6 +1137,95 @@ int config_parse_keep_configuration(
+                r = parse_sec(rvalue, &usec);
+                if (r < 0) {
+                        return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                       "Failed to parse [%s] %s=, ignoring assignment: %s", section, lvalue, rvalue);
```
Please align with `(`
```c
                        return log_syntax(unit, LOG_WARNING, filename, line, r,
                                          "Failed to parse [%s] %s=, ignoring assignment: %s",
                                          section, lvalue, rvalue);
```

yuwata commented on src/network/networkd-network.c:
```diff
@@ -1133,6 +1137,95 @@ int config_parse_keep_configuration(
+
+                if ((usec < MIN_EXPIRED_LEASE_LIFETIME_USEC || usec > MAX_EXPIRED_LEASE_LIFETIME_USEC) && !network_test_mode_enabled()) {
+                         return log_syntax(unit, LOG_WARNING, filename, line, 0,
+                                        "Extension lifetime value out of range, ignoring assignment: %s", rvalue);
```
Also here.

yuwata reviewed (changes requested):
Sorry for late review.

Please rearrange the commits, rather than stack fixes.

Also, for newer code, please use json format to save data, like dhcp-server does.

ozbenh commented on src/network/networkd-network-gperf.gperf:
```diff
@@ -303,6 +303,9 @@ DHCPv4.IPv6OnlyMode,                             config_parse_tristate,
 DHCPv4.RapidCommit,                              config_parse_tristate,                          0,                                      offsetof(Network, dhcp_use_rapid_commit)
+DHCPv4.DHCPClientPersistLeases,                  config_parse_dhcp_client_persist_leases,        0,                                      offsetof(Network, dhcp_client_persist_leases)
+DHCPv4.ExpiredLeaseExtensionLifetime,            config_parse_expired_lease_extension_options,   0,                                      offsetof(Network, expired_lease_extension_lifetime)
+DHCPv4.ExpiredLeaseExtensionFuzz,                config_parse_expired_lease_extension_options,   0,                                      offsetof(Network, expired_lease_extension_fuzz)
```
The idea here is to provide a bit of control on the "retry" of the clients, especially the fuzz factor to avoid having a massive herd of requests hammering the server who is just trying to get back online after a zone power failure.

rajibade commented:
Ack on using json format but what do you mean by stacking fixes here? Do you mean to separate the commits by the original commit dates instead of all together when doing rebase? or something else? and rearrange in what manner?

rajibade commented on src/libsystemd-network/sd-dhcp-lease.c:
```diff
@@ -1284,7 +1284,7 @@ int dhcp_lease_new(sd_dhcp_lease **ret) {
 }
 
-int dhcp_lease_save(sd_dhcp_lease *lease, const char *lease_file) {
+int dhcp_lease_save(sd_dhcp_lease *lease, int dir_fd, const char *lease_file) {
```
This function is the one you are referring to right?

rajibade replied:
Also if saving as json, I need to change the corresponding loading function too right?

rajibade committed:
network/dhcp-client: fix code format & use to new string table macro

rajibade committed:
network/dhcp-client: Use json format for client lease storage

Moved dhcp4 client lease loading and saving functions to use json format
to align with new server format. Fixed return error in client_initialize
from previous commit and updated fuzz_dhcp_client functio to use dir_fd
parameter from changed dhcp_lease_save

rajibade committed:
network/dhcp-client: cleanup funcs, error calls and unused header files

rajibade commented:
Hey @yuwata, Can I get a review on this? Its quite a large change, but it should move all the code towards saving as json. Appreciate the review on the original feature as well, and get this tagged for v261

rajibade commented:
I'll try to get any CI errors as they come

rajibade commented:
Oh here is a json output showing most of the exportable lease items and the format:

> root@myhost:~# cat /var/lib/systemd/network/dhcp-client-leases/enp0s2
> {"AddressString":"192.168.1.153","Address":[192,168,1,153],"NetmaskString":"255.255.255.0","Netmask":[255,255,255,0],"Router":[[192,168,1,1]],"Server_Address":[192,168,1,1],"Broadcast":[192,168,1,255],"MTU":1500,"T1":"13s","T2":"22s","Lifetime":"26s","Servers":{"DNS":[[1,1,1,1],[1,0,0,1]],"NTP":[[192,168,1,10],[192,168,1,11]],"SIP":[[192,168,1,20]],"POP3":[[192,168,1,30]],"SMTP":[[192,168,1,40]],"LPR":[[192,168,1,50]]},"DNR":[{"Priority":1,"Addresses":[[1,1,1,1]],"Port":853,"ServerName":"dns1.example.local","Transports":["dot"]},{"Priority":2,"Addresses":[[1,0,0,1]],"Port":853,"ServerName":"dns2.example.local","Transports":["dot"]},{"Priority":3,"Addresses":[[8,8,8,8]],"Port":443,"ServerName":"doh1.example.local","DoHPath":"/dns-query","Transports":["h2"]},{"Priority":4,"Addresses":[[8,8,4,4]],"Port":443,"ServerName":"doh2.example.local","DoHPath":"/dns-query","Transports":["h3"]}],"Domain_Name":"example.local","Domain_Search_List":["example.local","local"],"Hostname":"myhost","Root_Path":"/nfs/root","Static_Routes":[{"DestinationString":"192.168.10.0","Destination":[192,168,10,0],"GatewayString":"10.0.0.1","Gateway":[10,0,0,1],"DestinationPrefixLength":24},{"DestinationString":"10.0.0.0","Destination":[10,0,0,0],"GatewayString":"192.168.1.1","Gateway":[192,168,1,1],"DestinationPrefixLength":8}],"Classless_Routes":[{"DestinationString":"192.168.10.0","Destination":[192,168,10,0],"GatewayString":"10.0.0.1","Gateway":[10,0,0,1],"DestinationPrefixLength":24},{"DestinationString":"10.0.0.0","Destination":[10,0,0,0],"GatewayString":"10.0.0.1","Gateway":[10,0,0,1],"DestinationPrefixLength":8}],"Timezone":"UTC","Client_Id":[255,92,187,235,94,0,2,0,0,171,17,201,253,47,120,57,255,147,23],"Timestamp_realtime":"Wed 2026-04-22 19:53:58.113768 UTC","Vendor_Specific":"020568656c6c6f0104c0a80101","Private_Options":[{"Option":224,"Data":"736f6d6564617461"},{"Option":230,"Data":"00003039"},{"Option":249,"Data":"18c0a80a0a000001080ac0a80101"},{"Option":252,"Data":"687474703a2f2f777061642e6578616d706c652e6c6f63616c2f777061642e646174"}]}

rajibade commented:
@yuwata, I took at look at this draft of yours #41033, does this overlap with some of what is here? Can this be reviewed, so see if things need to be moved around?

EDIT: This too, #41796

rajibade commented:
From this: https://github.com/systemd/systemd/pull/41033/commits/0491af75d95eb8fba9846d4338994c683074af39
are you saying here to drop the saving of the lease?

rajibade commented:
Hey @yuwata. A few conflicts I'm holding off on here as it will require refactoring some of the code. I'll add documentation in a bit but can the logic in here be checked out?

yuwata commented:
Sorry for late response.

You may already know, now I am working on huge DHCP refactoring in #41033, which provides new serialization of DHCP lease in JSON format, and drops old code. Please use the new way.
Sorry for conflicting your work.

ozbenh commented:
> You may already know, now I am working on huge DHCP refactoring in #41033, which provides new serialization of DHCP lease in JSON format, and drops old code. Please use the new way. Sorry for conflicting your work.

Thanks for the reply ! That's ok, we missed that you were adding a new way, we thought you were just removing the old one. We can wait and adapt our work to your new code.

One thing we would want (this is the subject of customer issue we are dealing with) is to ensure that the complete DHCP response is saved (including the options networkd doesn't care about). Quite a few cases in the wild of people putting data in the DHCP reply and wanting to retrieve it from userspace :) (typically hijacking some of the netboot related options for example such as file name).

rajibade commented:
Appreciate the update @yuwata. I think there will be some benefit with us syncing together to get this sorted out, @ozbenh sent out an email earlier in the week, please let us know if this is possible. 

My recents commits already addressed saving the lease in JSON format in file(this is still possible with your new way correct?), not using the networkctl dump. From just parsing your draft PR, there is just a bit too much going on to pinpoint where to accurately make my changes. Plus there is the addition of https://github.com/systemd/systemd/pull/41780, which almost goes together. I don't want to prolong this PR by having to re-contextualise how everything works now. It'll be helpful to be collab so we stay on the same page. 

Let me know what you think. :)

rajibade committed:
man: document dhcpv4 persistent lease configuration

rajibade committed:
NEWS: include new DHCPv4 features

rajibade commented:
Added Docs. @yuwata

rajibade committed:
dhcp: Save/load with client internal fd

rajibade commented:
@yuwata can i get a review on this once your refactor is done? i'm not sure what new way you are referring to. does the serialization only affect the way i did the json building of the lease file? or is there much more difference?

rajibade commented:
Is there a cutoff to get be able to get this into v261 (can this be added to the milestone :-))? I'd like to know what I have to work with? @yuwata @bluca @poettering @whoever?
